src/dbnode/storage/throttler.go

```diff
@@ -0,0 +1,142 @@
+package storage
+
+import (
+	"sync"
+)
+
+// Throttler controls fair access to limited resources.
+type Throttler struct {
+	sync.Mutex
+
+	keyState map[string]*keyContext
+	// Each entry in the queue should be unique to avoid unfair access
+	keyQueue []string
+
+	globalCurrentWeight int
+	globalMaxWeight     int
+}
+
+type keyContext struct {
+	// Requests for weight to be granted which are waiting.
+	waiting []request
+	// Currently granted weight for this key.
+	currentWeight int
+}
+
+type request struct {
+	blockCh chan struct{}
+	weight  int
+}
+
+// Acquire blocks until the requested weight is granted to the specified key.
+func (t *Throttler) Acquire(key string, weight int) error {
+	blockCh, err := t.tryAcquire(key, weight)
+	if err != nil {
+		return err
+	}
+
+	if blockCh != nil {
+		<-blockCh
+	}
+
+	return nil
+}
+
+func (t *Throttler) tryAcquire(key string, weight int) (chan struct{}, error) {
+	t.Lock()
+	defer t.Unlock()
+
+	maxWeightPerKey := t.maxWeightPerKey()
+
+	currentKey, alreadyExists := t.keyState[key]
+	if !alreadyExists {
+		t.keyState[key] = &keyContext{
+			currentWeight: 0,
+			waiting:       make([]request, 0),
+		}
+	}
+
+	// If new weight would keep the key under the per-key limit then grant.
+	// Otherwise enqueue and block.
+	if currentKey.currentWeight+weight < maxWeightPerKey {
+		currentKey.currentWeight += weight
+		t.globalCurrentWeight += weight
```

Comment:
we don't enforce the global weight limit on `Acquire`, but only on attempts to acquire resources that get triggered from `Release`? seems weird.

Reply:
Yeah just an oversight. Addressed.